### PR TITLE
Port harvest to scripts and processes

### PR DESCRIPTION
**Pull request by YanaDePauw**
## Description
This PR contains a port from the harvest script to the new "Scripts and Processes" system in DSpace 7. 

Furthermore, the REST -h option has been improved to show more information about the different arguments instead of just a list of arguments. 

## Instructions for Reviewers
Verify that the harvest script can still be used as before from both the CLI and the REST: 
1. Use the -h option from both the CLI and REST and verify that they contain similar information about how to use the script.
2. Use the -s option as specified in the help option to setup a collection for harvesting. 
```
POST {{url}}/api/system/scripts/harvest/processes
properties: 
[
  {
    "name": "-s"
  },
  {
    "name": "-c",
    "value": "{{collection-handle}}"
  },
  {
    "name": "-a",
    "value": "{{oai-server-address}}"
  },
  {
    "name": "-i",
    "value": "com_2384_294335"
  },
  {
    "name": "-m",
    "value": "dc"
} ,
  {
    "name": "-t",
    "value": "1"
  }
]
```
3. Use the -r option as specified in the help option to run the harvest on the previously configured collection. 
```
POST  {{url}}/api/system/scripts/harvest/processes
properties: 
[
  {
    "name": "-r"
  },
  {
    "name": "-c",
    "value": "{{collection-handle}}"
  },
  {
    "name": "-e",
    "value": "{{user-email}}"
  }
]
```
4. Verify that items are harvested in the collection and that "api/core/collections/{collection-uuid}/harvester" contains the harvest config you set up in step 2 and the harvest start time from step 3. 
5. Use the -p option as specified in the help option to purge the collection from the harvested items.
```
POST  {{url}}/api/system/scripts/harvest/processes
properties: 
[
  {
    "name": "-p"
  },
  {
    "name": "-c",
    "value": "{{collection-handle}}"
  },
  {
    "name": "-e",
    "value": "{{user-email}}"
  }
]
```

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/harvest/Harvest.java [hunk 1]
```diff
@@ -36,221 +33,224 @@
...
+                handler.logError("A target collection and eperson must be provided (run with -h flag for details)");
+                throw new UnsupportedOperationException("A target collection and eperson must be provided");
             }
 
-            harvester.purgeCollection(collection, eperson);
+            purgeCollection(context, collection);
             context.complete();
 
             //TODO: implement this... remove all items and remember to unset "last-harvested" settings
+        } else if ("reimport".equals(command)) {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/harvest/Harvest.java:202]: Can the TODO above this line be removed now?  It looks like you've implemented "reimport"

File: dspace-api/src/main/java/org/dspace/app/harvest/HarvestCli.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import java.sql.SQLException;
+
+import org.apache.commons.cli.ParseException;
+import org.dspace.core.Context;
+import org.dspace.eperson.EPerson;
+
+/**
+ * Test class for harvested collections.
+ *
+ * @author Alexey Maslov
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/harvest/HarvestCli.java:19]: Silly thing, but it stuck out immediately.  Is this author tag correct?  If not, either update it or just remove it, as we don't require author tags

File: dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    public Options getOptions() {
+        Options options = new Options();
+        options.addOption("p", "purge", false, "delete all items in the collection");
+        options.addOption("r", "run", false, "run the standard harvest procedure");
+        options.addOption("g", "ping", false, "test the OAI server and set");
+        options.addOption("s", "setup", false, "Set the collection up for harvesting");
+        options.addOption("S", "start", false, "start the harvest loop");
+        options.addOption("R", "reset", false, "reset harvest status on all collections");
+        options.addOption("P", "purge", false, "purge all harvestable collections");
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java:51]: This and the `-p` option both use `--purge`...  Which one will be triggered if you run `--purge`? Maybe we should update this one to be `--purgeCollections` or similar?

File: dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        options.addOption("o", "reimport", false, "reimport all items in the collection");
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java:52]: We should add a note to the description here that `reimport` seems to be the equivalent of `-p -r` (if I'm reading the code correctly). So, this should warn that we are purging all contents (i.e. deleting existing items) and re-importing from scratch.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@YanaDePauw : Gave this a review today and it looks good overall.  I only have some minor suggestions / feedback to be addressed (see below).  I also tested this from the commandline & it works well.  However, I had some issues in testing from the UI, but I'll report those in my feedback on your UI PR https://github.com/DSpace/dspace-angular/pull/1328

File: dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java (hunk 3 above)
Comment (LucaGiamminonni) [reply dspace-api/src/main/java/org/dspace/app/harvest/HarvestScriptConfiguration.java:51]: I agree with @tdonohue, this ---purge should be changed to --purgeCollections

**Review by LucaGiamminonni — APPROVED** [review]
Hi @YanaDePauw, I had a look at the changes and they look good to me. The only thing I would change is the purge parameter as recommended by @tdonohue in the specific comment. I tried to start harvest from both UI and CLI and I had no problems.

**Review by tdonohue — APPROVED** [review]
👍 My feedback on this PR has all been addressed.  That said, I'm going to wait to merge this until the Angular UI PR is ready...as there's a bug in the Processes UI that may or may not require backend changes. See my review there: https://github.com/DSpace/dspace-angular/pull/1328#pullrequestreview-775425176

**Review by tdonohue — APPROVED** [review]
👍 Retested today with the new minor updates, and the param fixes did resolve the odd UI bug I was seeing.  Thanks @YanaDePauw !

**Pull request merged**